pkg/internal/stream/stream.go

```diff
@@ -0,0 +1,420 @@
+				timerCtx, timerCancel = context.WithCancel(ctx)
+				lock.Unlock()
+				timer := clk.Timer(debounceTime)
+				go func() {
+					select {
+					case <-timerCtx.Done():
+						if !timer.Stop() {
+							<-timer.C
+						}
+					case <-timer.C:
+						emit()
+					}
+				}()
+			},
+			onError: func(err error) {
+				emit()
+				subscriber.Error(err)
+			},
+			onComplete: func() {
+				emit()
+				subscriber.Complete()
+			},
+		}
+		source.Subscribe(debounceSubscriber)
+	})
+}
+
+// Map transforms values on a channel based on the passed in transformation function
+func Map[T any, U any](stream types.Stream[T], mapFn func(T) (U, error)) types.Stream[U] {
+	return lift(stream, func(source types.Stream[T], subscriber types.StreamSubscriber[U]) {
+		mapSubscriber := operatorSubscriber(subscriber, func(t T) {
+			next, err := mapFn(t)
+			if err != nil {
+				subscriber.Error(err)
+				return
+			}
+			subscriber.Next(next)
+		})
+		source.Subscribe(mapSubscriber)
+	})
+}
+
+// Filter filters values on a channel based on the passed in function -- if the function returns true, the value
+// is emmitted, if not, the value is not emitted
+func Filter[T any](source types.Stream[T], include func(T) bool) types.Stream[T] {
+	return lift(source, func(source types.Stream[T], subscriber types.StreamSubscriber[T]) {
+		filterSubscriber := operatorSubscriber(subscriber, func(t T) {
+			if include(t) {
+				subscriber.Next(t)
+			}
+		})
+		source.Subscribe(filterSubscriber)
+	})
+}
+
+func Take[T any](source types.Stream[T], count uint64) types.Stream[T] {
+	return lift(source, func(source types.Stream[T], subscriber types.StreamSubscriber[T]) {
+		seen := uint64(0)
+		takeSubscriber := operatorSubscriber(subscriber, func(t T) {
+			seen++
+			if seen <= count {
+				subscriber.Next(t)
+				if count <= seen {
+					subscriber.Complete()
+				}
+			}
+		})
+		source.Subscribe(takeSubscriber)
+	})
+}
+
+func TakeWhile[T any](source types.Stream[T], while func(T) bool, inclusive bool) types.Stream[T] {
+	return lift(source, func(source types.Stream[T], subscriber types.StreamSubscriber[T]) {
+		takeSubscriber := operatorSubscriber(subscriber, func(t T) {
+			if while(t) {
+				subscriber.Next(t)
+				return
+			}
+			if inclusive {
+				subscriber.Next(t)
+			}
+			subscriber.Complete()
+		})
+		source.Subscribe(takeSubscriber)
+	})
+}
+
+func Last[T any](source types.Stream[T]) types.Stream[T] {
+	return lift(source, func(source types.Stream[T], subscriber types.StreamSubscriber[T]) {
+		var last T
+		var hasLast bool
+		lastSubscriber := &internalSubscriber[T]{
+			onNext: func(t T) {
+				hasLast = true
+				last = t
+			},
+			onError: func(err error) {
+				subscriber.Error(err)
+			},
+			onComplete: func() {
+				if hasLast {
+					subscriber.Next(last)
+				}
+				subscriber.Complete()
+			},
+		}
+		source.Subscribe(lastSubscriber)
+	})
+}
+
+func Scan[T any, U any](stream types.Stream[T], scanFn func(U, T) (U, error), start U) types.Stream[U] {
```

Comment:
are you seeing a use for scan, take, etc. that warrants including them? or are you just filling out a feature set to make it feel generically complete?

Reply:
well I'm using all of them in some place. see how the converter for async retrieval works.